File: backend/app/app/models/project.py
from typing import TYPE_CHECKING

import datetime
from sqlalchemy import Column, String, Integer, ForeignKey, DateTime, func
from sqlalchemy.orm import relationship

from app.db.base_class import Base
from app.models.project_worker import ProjectWorker

if TYPE_CHECKING:
    pass


class Project(Base):
    __tablename__ = 'project'

    id = Column(Integer, primary_key=True)
    user_id = Column(Integer, ForeignKey('user.id'), index=True, nullable=False)
    name = Column(String)
    address = Column(String)
    description = Column(String)
    created_dt = Column(DateTime, default=datetime.datetime.utcnow)
    modified_dt = Column(DateTime, default=datetime.datetime.utcnow, onupdate=func.current_timestamp())
    removed_dt = Column(DateTime)

    workers = relationship('User', secondary=ProjectWorker.__tablename__, back_populates="project_worker")
    project_worker = relationship('ProjectWorker')
    client = relationship('User')

File: frontend/src/interfaces/Project.ts
import {IWorkerProfile, IWorkerProfileAdmin, IProjectWorker, IProjectWorkerAdminCreateUpdate} from './index';

export interface IProject {
  id: number;
  name: string;
  user_id: number;
  address: string;
  description: string;
  created_dt: Date;
  modified_dt: Date;
  removed_dt: Date | null;
  project_worker: IProjectWorker[]
}

export interface IProjectAdminCreateUpdate {
  id?: number;
  user_id?: number;
  name?: string;
  address?: string;
  description?: string;
  project_worker?: IProjectWorkerAdminCreateUpdate[]
}

export interface IProjectCreate {
  user_id: number;
  name: string;
  address: string;
  description?: string;
  workers: IWorkerProfile[];
}

export interface IProjectForProjectWorker {
  user_id: number;
  name: string;
  address: string;
  description?: string;
  workers: IWorkerProfileAdmin[];
}

export interface IProjectLimited {
  id: number;
  user_id: number;
  name: string;
  address: string;
  description?: string;
}